### PR TITLE
Fix Cargo pre-1.0 semantic versioning classification (0.y.z minor updates as major)

### PR DESCRIPTION
**Pull request by thavaahariharangit**
### What are you trying to accomplish?

Fixes Cargo dependency update type classification to properly handle pre-1.0 semantic versioning rules where changes in the leftmost non-zero component are considered breaking changes.

Previously, Dependabot incorrectly classified updates like `annotate-snippets` from `0.11.5` to `0.12.5` as "minor" updates instead of "major" updates. According to Cargo's semantic versioning conventions:

- For `0.y.z` versions: changes in `y` are considered major/breaking changes
- For `0.0.z` versions: changes in `z` are considered major/breaking changes
- Only the leftmost non-zero component determines compatibility

Ref: https://github.com/dependabot/dependabot-core/issues/13337

### Anything you want to highlight for special attention from reviewers?

1. Enhanced Cargo::Version class
- Added `update_type` class method that correctly classifies version changes according to Cargo's pre-1.0 semver rules
- Override `ignored_patch_versions` and `ignored_minor_versions` to respect Cargo's compatibility rules
- Added comprehensive helper methods for version classification

2. Updated Group Update Logic
- Modified `semver_rules_allow_grouping?` in group update creation to use Cargo-specific version classification
- Added proper handling for Cargo package manager in dependency grouping scenarios

3. Fixed Feature Error Handling
- Enhanced error handling for Cargo feature mismatch errors (both singular "does not have that feature" and plural "does not have these features")
- Improved error message handling for better debugging

4. Comprehensive Test Coverage
- Added extensive RSpec tests for all new version classification logic
- Added tests for pre-1.0 version scenarios including `0.y.z` and `0.0.z` cases
- Added integration tests for dependency grouping scenarios

5. Examples of Fixed Behavior

| From Version | To Version | Before | After | Reasoning |
|--------------|------------|--------|-------|-----------|
| `0.11.5` | `0.12.5` | minor | **major** | In `0.y.z`, minor changes are breaking |
| `0.0.3` | `0.0.4` | patch | **major** | In `0.0.z`, patch changes are breaking |
| `0.2.3` | `0.2.4` | patch | patch | In `0.y.z`, patch changes are compatible |
| `1.2.3` | `1.3.0` | minor | minor | Standard semver for 1.0.0+ versions |

### How will you know you've accomplished your goal?

Testing
- All existing Cargo tests pass
- New comprehensive test suite covers edge cases and integration scenarios
- Verified with real-world examples like `annotate-snippets` updates

Ex: Ran cli against given repo: https://github.com/VorpalBlade/paketkoll
```
updater | 2025/10/21 12:44:54 INFO Finished job processing
updater | 2025/10/21 12:44:54 INFO Results:
updater | +-----------------------------------------------------------------------------------------------+
updater | |                              Changes to Dependabot Pull Requests                              |
updater | +---------+-------------------------------------------------------------------------------------+
updater | | created | annotate-snippets ( from 0.11.5 to 0.12.5 ), rune-modules ( from 0.13.4 to 0.14.1 ) |
updater | | created | clap ( from 4.5.49 to 4.5.50 ), clap_builder ( from 4.5.49 to 4.5.50 )              |
updater | +---------+-------------------------------------------------------------------------------------+
```

References
- [Cargo SemVer Compatibility Guide](https://doc.rust-lang.org/cargo/reference/resolver.html#semver-compatibility)
- [Rust RFC: Semantic Versioning](https://github.com/rust-lang/rfcs/blob/master/text/1105-api-evolution.md)

Fixes issues where Cargo pre-1.0 dependencies were incorrectly classified for dependency grouping and update strategies.

### Checklist

- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

### CODE REVIEW

File: cargo/lib/dependabot/cargo/update_checker/version_resolver.rb [hunk 1]
```diff
@@ -238,75 +238,100 @@ def check_rust_workspace_root
...
             next unless error.message.match?(regex)
 
             dependency_url = T.must(T.must(error.message.match(regex)).named_captures.fetch("url")).split(/[#?]/).first
-            # Fix: Wrap dependency_url in T.must since split().first can return nil
             raise Dependabot::GitDependencyReferenceNotFound, T.must(dependency_url)
           end
+        end
 
+        sig { params(error: StandardError).returns(T::Boolean) }
+        def workspace_error_should_return_nil?(error)
```
Comment (markhallen) [inline cargo/lib/dependabot/cargo/update_checker/version_resolver.rb]: nit, this method name is a bit confusing. I don't have a better name, but if you can share a bit more context maybe I can help you to come up with one.

**Review by markhallen — APPROVED** [review]
